# Logistic Regression

Mainly for 2 classify between 0 or 1

The p value equal to e^(corarelation of vlaue with other dataset
                    OR
 The p value equal to e^(mx+c(i.e regression)
 sigmoid eq in terms of is p=(e^(mx+c))/(1+e^(mx+c))

# Importing the pakages and setting up the enviroment

Importing all pakages in the first is good Practice

In [2]:
using Pkg
Pkg.add("ClassImbalance")
Pkg.add("ROCAnalysis")
Pkg.add("StatsBase")

  Resolving package versions...
  Installed UnPack ───────────────── v1.0.2
  Installed KernelDensity ────────── v0.5.1
  Installed Lathe ────────────────── v0.0.9
  Installed ClassImbalance ───────── v0.8.7
  Installed ForwardDiff ──────────── v0.10.18
  Installed NLSolversBase ────────── v7.5.0
  Installed ArrayInterface ───────── v2.14.17
  Installed DataFrames ───────────── v0.20.2
  Installed StatsModels ──────────── v0.6.21
  Installed SpecialFunctions ─────── v0.10.3
  Installed CommonSubexpressions ─── v0.3.0
  Installed Grisu ────────────────── v1.0.2
  Installed DataStructures ───────── v0.17.20
  Installed CategoricalArrays ────── v0.7.7
  Installed ColorTypes ───────────── v0.10.12
  Installed DiffEqDiffTools ──────── v1.7.0
  Installed FillArrays ───────────── v0.8.14
  Installed DiffResults ──────────── v1.0.3
  Installed LineSearches ─────────── v7.1.1
  Installed PDMats ───────────────── v0.9.12
  Installed PooledArrays ─────────── v0.5.3
  Installed Calculus ──────────

In [3]:
using CSV
using DataFrames
using Plots
using Lathe
using GLM
using StatsBase
using MLBase
using ClassImbalance
using ROCAnalysis
# Enable printing of 1000 columns
ENV["COLUMNS"] = 1000

┌ Info: Precompiling CSV [336ed68f-0bac-5ca0-87d4-7b16caf5d00b]
└ @ Base loading.jl:1278
┌ Info: Precompiling DataFrames [a93c6f00-e57d-5684-b7b6-d8193f3e46c0]
└ @ Base loading.jl:1278
┌ Info: Precompiling Plots [91a5bcdd-55d7-5caf-9e0b-520d859cae80]
└ @ Base loading.jl:1278
┌ Info: Precompiling Lathe [38d8eb38-e7b1-11e9-0012-376b6c802672]
└ @ Base loading.jl:1278
┌ Info: Precompiling GLM [38e38edf-8417-5370-95a0-9cbb8c7f171a]
└ @ Base loading.jl:1278
┌ Info: Precompiling MLBase [f0e99cf1-93fa-52ec-9ecc-5026115318e0]
└ @ Base loading.jl:1278
┌ Info: Precompiling ClassImbalance [04a18a73-7590-580c-b363-eeca0919eb2a]
└ @ Base loading.jl:1278
┌ Info: Precompiling ROCAnalysis [f535d66d-59bb-5153-8d2b-ef0a426c6aff]
└ @ Base loading.jl:1278


1000

# Loading the Data in the Notebook

In [9]:
#Read the file using CSV.FILE and convert it to DataFrame
df=DataFrame(CSV.File("Churn_Modelling.csv"))
first(df,5)

5×14 DataFrame
│ Row │ RowNumber │ CustomerId │ Surname  │ CreditScore │ Geography │ Gender │ Age   │ Tenure │ Balance   │ NumOfProducts │ HasCrCard │ IsActiveMember │ EstimatedSalary │ Exited │
│     │ Int64     │ Int64      │ String   │ Int64       │ String    │ String │ Int64 │ Int64  │ Float64   │ Int64         │ Int64     │ Int64          │ Float64         │ Int64  │
├─────┼───────────┼────────────┼──────────┼─────────────┼───────────┼────────┼───────┼────────┼───────────┼───────────────┼───────────┼────────────────┼─────────────────┼────────┤
│ 1   │ 1         │ 15634602   │ Hargrave │ 619         │ France    │ Female │ 42    │ 2      │ 0.0       │ 1             │ 1         │ 1              │ 1.01349e5       │ 1      │
│ 2   │ 2         │ 15647311   │ Hill     │ 608         │ Spain     │ Female │ 41    │ 1      │ 83807.9   │ 1             │ 0         │ 1              │ 1.12543e5       │ 0      │
│ 3   │ 3         │ 15619304   │ Onio     │ 502         │ France    │ Female │ 42    │ 8      │ 1.59661e5 │ 3             │ 1         │ 0              │ 1.13932e5       │ 1      │
│ 4   │ 4         │ 15701354   │ Boni     │ 699         │ France    │ Female │ 39    │ 1      │ 0.0       │ 2             │ 0         │ 0              │ 93826.6         │ 0      │
│ 5   │ 5         │ 15737888   │ Mitchell │ 850         │ Spain     │ Female │ 43    │ 2      │ 1.25511e5 │ 1             │ 1         │ 1              │ 79084.1         │ 0      │

# Summary of the DataFrAme

In [10]:
#Summary of dataframe
println(size(df))
describe(df)

(10000, 14)


14×8 DataFrame
│ Row │ variable        │ mean      │ min      │ median    │ max       │ nunique │ nmissing │ eltype   │
│     │ Symbol          │ Union…    │ Any      │ Union…    │ Any       │ Union…  │ Nothing  │ DataType │
├─────┼─────────────────┼───────────┼──────────┼───────────┼───────────┼─────────┼──────────┼──────────┤
│ 1   │ RowNumber       │ 5000.5    │ 1        │ 5000.5    │ 10000     │         │          │ Int64    │
│ 2   │ CustomerId      │ 1.56909e7 │ 15565701 │ 1.56907e7 │ 15815690  │         │          │ Int64    │
│ 3   │ Surname         │           │ Abazu    │           │ Zuyeva    │ 2932    │          │ String   │
│ 4   │ CreditScore     │ 650.529   │ 350      │ 652.0     │ 850       │         │          │ Int64    │
│ 5   │ Geography       │           │ France   │           │ Spain     │ 3       │          │ String   │
│ 6   │ Gender          │           │ Female   │           │ Male      │ 2       │          │ String   │
│ 7   │ Age             │ 38.9218   │ 18       │ 37.0      │ 92        │         │          │ Int64    │
│ 8   │ Tenure          │ 5.0128    │ 0        │ 5.0       │ 10        │         │          │ Int64    │
│ 9   │ Balance         │ 76485.9   │ 0.0      │ 97198.5   │ 2.50898e5 │         │          │ Float64  │
│ 10  │ NumOfProducts   │ 1.5302    │ 1        │ 1.0       │ 4         │         │          │ Int64    │
│ 11  │ HasCrCard       │ 0.7055    │ 0        │ 1.0       │ 1         │         │          │ Int64    │
│ 12  │ IsActiveMember  │ 0.5151    │ 0        │ 1.0       │ 1         │         │          │ Int64    │
│ 13  │ EstimatedSalary │ 1.0009e5  │ 11.58    │ 1.00194e5 │ 1.99992e5 │         │          │ Float64  │
│ 14  │ Exited          │ 0.2037    │ 0        │ 0.0       │ 1         │         │          │ Int64    │

# Checking the names of all the columns in the Dataframe

In [19]:
#check the columns names
names(df)

14-element Array{Symbol,1}:
 :RowNumber
 :CustomerId
 :Surname
 :CreditScore
 :Geography
 :Gender
 :Age
 :Tenure
 :Balance
 :NumOfProducts
 :HasCrCard
 :IsActiveMember
 :EstimatedSalary
 :Exited

# Checking the no of response variable

In [20]:
#count the classes
countmap(df.Exited)
#the countmap function counts the no of unique entries in the dataframe's column

Dict{String,Int64} with 3 entries:
  "Germany" => 2509
  "Spain"   => 2477
  "France"  => 5014

# One Hot Encoding

3 categorical columns are present in the dataset.These columns need to be handled.The surname column is having 2932 unique values,so it s not good idea to encode it.Rather lets drop it and encode Gender and Geography column

In [23]:
Lathe.preprocess.OneHotEncode(df,:Geography)
Lathe.preprocess.OneHotEncode(df,:Gender)
select!(df,Not([:RowNumber,:CustomerId,:Surname,:Geography,:Gender,:Male]))#Removing Columns that are unrevelant to our model RowNumber,:CustomerId,:Surname,:Geography,:Gender(Male) male is remove due to dummy trap to decrease parsing 
first(df,5)

LoadError: ArgumentError: column name :Geography not found in the data frame

# Train Test Splitting

In [24]:
#Train Test Split
using Lathe.preprocess: TrainTestSplit
train,test=TrainTestSplit(df,.75)

(7527×13 DataFrame
│ Row  │ CreditScore │ Age   │ Tenure │ Balance   │ NumOfProducts │ HasCrCard │ IsActiveMember │ EstimatedSalary │ Exited │ France │ Spain │ Germany │ Female │
│      │ Int64       │ Int64 │ Int64  │ Float64   │ Int64         │ Int64     │ Int64          │ Float64         │ Int64  │ Bool   │ Bool  │ Bool    │ Bool   │
├──────┼─────────────┼───────┼────────┼───────────┼───────────────┼───────────┼────────────────┼─────────────────┼────────┼────────┼───────┼─────────┼────────┤
│ 1    │ 619         │ 42    │ 2      │ 0.0       │ 1             │ 1         │ 1              │ 1.01349e5       │ 1      │ 1      │ 0     │ 0       │ 1      │
│ 2    │ 608         │ 41    │ 1      │ 83807.9   │ 1             │ 0         │ 1              │ 1.12543e5       │ 0      │ 0      │ 1     │ 0       │ 1      │
│ 3    │ 502         │ 42    │ 8      │ 1.59661e5 │ 3             │ 1         │ 0              │ 1.13932e5       │ 1      │ 1      │ 0     │ 0       │ 1      │
│ 4    │ 699         

# Model Building

In [27]:
fm=@formula(Exited~ CreditScore+Age+Tenure+Balance+NumOfProducts+HasCrCard+IsActiveMember+EstimatedSalary+Female+France +Spain)#instead of gender in column we have use Female(which was form due to One Hot Encoding,and intead of geography we have used spain & france
logit=glm(fm,train,Binomial(),ProbitLink())

StatsModels.TableRegressionModel{GeneralizedLinearModel{GLM.GlmResp{Array{Float64,1},Binomial{Float64},ProbitLink},GLM.DensePredChol{Float64,LinearAlgebra.Cholesky{Float64,Array{Float64,2}}}},Array{Float64,2}}

Exited ~ 1 + CreditScore + Age + Tenure + Balance + NumOfProducts + HasCrCard + IsActiveMember + EstimatedSalary + Female + France + Spain

Coefficients:
────────────────────────────────────────────────────────────────────────────────────────
                        Coef.   Std. Error       z  Pr(>|z|)     Lower 95%     Upper 95%
────────────────────────────────────────────────────────────────────────────────────────
(Intercept)      -1.72061      0.167238     -10.29    <1e-24  -2.04839      -1.39283
CreditScore      -0.000487487  0.000183525   -2.66    0.0079  -0.000847189  -0.000127784
Age               0.040547     0.00167397    24.22    <1e-99   0.037266      0.0438279
Tenure           -0.010789     0.00618103    -1.75    0.0809  -0.0229036     0.00132557
Balance           1

# Model Prediction and Evaluation


In [28]:
#Predict the target variable on test data
prediction= predict(logit,test)

2473-element Array{Union{Missing, Float64},1}:
 0.2314743258265768
 0.12891749180692894
 0.11508156437215356
 0.04777662920488347
 0.20374271756171855
 0.4552710621369863
 0.13358373164827794
 0.3985205353090804
 0.14290667142931793
 0.28133224514701277
 0.1824076664750081
 0.5834201807526846
 0.13829946112571062
 ⋮
 0.05976754124083309
 0.4308308352527016
 0.1307437922526073
 0.18393759441317004
 0.06003369128928001
 0.25076038509762216
 0.015104896567245925
 0.13662574747856537
 0.5175550454854396
 0.12424395331689639
 0.10733415444338391
 0.1287563089433761

In [ ]:
 The prediction of glm